Frameworks/CoreGraphics/DWriteFontBinaryDataCollectionLoader.mm

```diff
@@ -0,0 +1,170 @@
+                                                                               CFArrayGetValueAtIndex(m_fontDatas.get(), m_location))));
+            RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(loader.Get()));
+
+            int unused;
+            RETURN_IF_FAILED(dwriteFactory->CreateCustomFontFileReference(&unused, sizeof(unused), loader.Get(), fontFile));
+            m_previouslyCreatedFiles->emplace_back(*fontFile);
+        }
+
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE MoveNext(_Out_ BOOL* hasCurrentFile) {
+        RETURN_HR_IF_NULL(E_POINTER, hasCurrentFile);
+        *hasCurrentFile = ++m_location < CFArrayGetCount(m_fontDatas.get());
+        return S_OK;
+    }
+
+private:
+    int m_location = -1;
+
+    woc::unique_cf<CFArrayRef> m_fontDatas;
+    std::vector<ComPtr<IDWriteFontFile>>* m_previouslyCreatedFiles;
+};
+
+// DWriteFontBinaryDataCollectionLoader member functions
+
+HRESULT DWriteFontBinaryDataCollectionLoader::DWriteFontBinaryDataCollectionLoader::RuntimeClassInitialize() {
+    m_fontDatas.reset(CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks));
+    m_fontDatasSet.reset(CFSetCreateMutable(nullptr, 0, &kCFTypeSetCallBacks));
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE DWriteFontBinaryDataCollectionLoader::CreateEnumeratorFromKey(_In_ IDWriteFactory* factory,
+                                                                                        _In_ const void* collectionKey,
+                                                                                        _In_ UINT32 collectionKeySize,
+                                                                                        _Out_ IDWriteFontFileEnumerator** enumerator) {
+    MakeAndInitialize<DWriteFontFileEnumerator>(enumerator, m_fontDatas.get(), &m_previouslyCreatedFiles);
+    return S_OK;
+}
+
+HRESULT DWriteFontBinaryDataCollectionLoader::AddDatas(CFArrayRef fontDatas, CFArrayRef* errors) {
+    CFMutableArrayRef outErrors = nil;
+    HRESULT ret = S_OK;
+    if (errors) {
+        outErrors = CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks);
+        *errors = outErrors;
+    }
+
+    CFIndex count = CFArrayGetCount(fontDatas);
+    for (CFIndex i = 0; i < count; ++i) {
+        CFDataRef data = static_cast<CFDataRef>(CFArrayGetValueAtIndex(fontDatas, i));
+        if (data) {
+            if (CFSetContainsValue(m_fontDatasSet.get(), data)) {
+                __AppendErrorIfExists(outErrors, kCTFontManagerErrorAlreadyRegistered);
+                ret = S_FALSE;
+            } else {
+                CFArrayAppendValue(m_fontDatas.get(), data);
+                CFSetAddValue(m_fontDatasSet.get(), data);
+                __AppendNullptrIfExists(outErrors);
+            }
+        } else {
+            __AppendErrorIfExists(outErrors, kCTFontManagerErrorInvalidFontData);
+            ret = S_FALSE;
```

Comment:
Mm, pushing back on this. I'm not convinced that a single iteration failing here should fail the whole function, and I don't think that RETURN_IF_FAILED should be tripped in such a case. S_FALSE has a connotation of 'partial success', and we have the errors out-array if we need more specific error information anyway.

---
In reply to: [90743736](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90743736) [](ancestors = 90743736)